examples/axios-typescript/libs/useRequest.ts

```diff
@@ -0,0 +1,52 @@
+import useSWR, { ConfigInterface, responseInterface } from 'swr'
+import axios, { AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios'
+import { useMemo } from 'react'
+
+export type GetRequest = AxiosRequestConfig | null
+
+interface Return<Data, Error>
+  extends Pick<
+    responseInterface<AxiosResponse<Data>, AxiosError<Error>>,
+    'isValidating' | 'revalidate' | 'error'
+  > {
+  data: Data | undefined
+  response: AxiosResponse<Data> | undefined
+}
+
+export interface Config<Data = unknown, Error = unknown>
+  extends Omit<
+    ConfigInterface<AxiosResponse<Data>, AxiosError<Error>>,
+    'initialData'
+  > {
+  initialData?: Data
+}
+
+export default function useRequest<Data = unknown, Error = unknown>(
+  request: GetRequest,
+  { initialData, ...config }: Config<Data, Error> = {}
+): Return<Data, Error> {
+  const { data: response, error, isValidating, revalidate } = useSWR<
+    AxiosResponse<Data>,
+    AxiosError<Error>
+  >(request && JSON.stringify(request), () => axios(request || {}), {
```

Comment:
I think we should add a comment here and mention it in the README (and the documentation in the future).
`JSON.stringify` is a hack and not the recommended way because:
1. it's unnecessary and slow (runs on every render)
2. it's not stable

2 means it can produce different keys for the same request. e.g.:
`JSON.stringify({ params: { a: 1, b: 2 } }) // -> {"params":{"a":1,"b":2}}`
`JSON.stringify({ params: { b: 2, a: 1 } }) // -> {"params":{"b":2,"a":1}}`

I think this will be better:
```js
useRequest(request, deps = [], config = {}) {
  useSWR(deps, () => axios(request || {}), ...)
  ...
}

// to use it:

useRequest({
  url: '/api/data',
  params: { id }
}, ['/api/data', id])

useRequest({
  url: '/api/data'
}, ['/api/data'])
```

Just like React's `useMemo`, `useEffect`, and `useCallback`.

But it's not an anti-pattern or "mistake" to use `JSON.stringify` here. I'm good with using it in this specific example and sometimes it is easier to use and working well for small apps indeed. That's why I think we need a comment in the example's code to clarify that.

Reply:
I like your suggestion. I'll try to update it with that 🙂

Reply:
@quietshu How does `useSWR` react if there's no key? Or if you pass it an empty array? Will it return the cached version of a different call that also was created with an empty array? Or will those arrays be different arrays, and if so, how would it be cached if the array next render "changes" to a new empty array? 🤔

Reply:
@svish SWR will "pause" the data fetching (not fetching anything and just return `undefined` as the data) if the key is an empty string, array or just null.

Reply:
So it wouldn't really work then, because you'd need something more to use as a key. An empty deps array, in the case of a query with no parameter, would be ignored, and an array with for example an id of 2, would potentially be mixed up with a different query which also had an id of 2 (for example a post and a user with the same id). 😕

Reply:
Yeah that's why I also put the `url` in the array: `['/api/data', id]`. So it won't be mixed up.

Reply:
Ooh, right, missed that. Thanks, and sorry for being a bother 😛

Reply:
No worries : )

Reply:
@quietshu Hm, after looking at it more, I think it's easier to just use the `JSON.stringify`, because it's difficult to know exactly how much in the request object one should need to include in the array. And from what I can understand, `JSON.stringify` is fairly fast, and it should also be stable as long as the object is created the same way each time.

Would probably be good to add a comment about that though, as you mention, but not sure how it should be written. Maybe one of you guys could try to author something along the lines of what you think it should say, and I can add it to the code? 🤔

Reply:
@Svish No worries, I think it's fine for now.

Reply:
@Svish Hi. With a current approach, I stuck with an issue when I need to mutate some request but I don't know its params cause they are too high in a tree of components and there is no sense to put it down to children.

Reply:
> @Svish Hi. With a current approach, I stuck with an issue when I need to mutate some request but I don't know its params cause they are too high in a tree of components and there is no sense to put it down to children.

Well, then you just need to do it differently then 😛 I'm currently only doing fetching with this library, so haven't done anything with mutate or such yet.

Reply:
Seems like there is a sense to pay attention to the "apollo-client" implementation of working with cache (__typename and etc) cc @quietshu